File: test/samples/floats/floats.lua
-- f (number): a float
-- g (number): a float, greater than f
-- point (table: {"x": number, "y": number, "z": number}): some coordinates
-- n (number): a number
-- float_list (table: array[number]): a list of floats
-- other_list (table: array[number]): a list of floats
-- inlined (table: array[{"integer": number, "char": string, "float": number}]): some inlined structs
-- multiline (table: {"integer 2": number, "string": string, "float 2": number}): a multiline struct
function floats(f, g, point, n, float_list, other_list, inlined, multiline)
    -- TODO Parsing is often easy, reprint mode is harder
end

local f = tonumber(io.read())
local g = tonumber(io.read())
local i = {string.match(io.read(), "(-?%g+) (-?%g+) (-?%g+)")}
local point = {x = tonumber(i[1]), y = tonumber(i[2]), z = tonumber(i[3])}
local n = tonumber(io.read())
local float_list = {}
for i in string.gmatch(io.read(), "-?%g+") do
    table.insert(float_list, tonumber(i))
end
local other_list = {}
for i in string.gmatch(io.read(), "-?%g+") do
    table.insert(other_list, tonumber(i))
end
local inlined = {}
for i = 1, 3 do
    local j = {string.match(io.read(), "(-?%d+) (%S) (-?%g+)")}
    inlined[i] = {integer = tonumber(j[1]), char = j[2], float = tonumber(j[3])}
end
local multiline = {}
multiline["integer 2"] = tonumber(io.read())
multiline["string"] = io.read()
multiline["float 2"] = tonumber(io.read())

floats(f, g, point, n, float_list, other_list, inlined, multiline)

File: test/samples/floats/floats.pas
program Floats;

type
    { Represents coordinates }
    Coordinates = record
        X: double; { X }
        Y: double; { Y }
        Z: double; { Z }
    end;

    { Mix of fields that go on one line }
    InlinedMix = record
        Integer_: longint; { an integer }
        Char_: char; { a char }
        Float: double; { a float }
    end;

    { a struct of chars }
    MultilineMix = record
        Integer2: longint; { an other integer }
        String_: AnsiString; { a string of size 5 }
        Float2: double; { an other float }
    end;

    T_FloatList = array of double;
    T_OtherList = array of double;
    T_Inlined = array of InlinedMix;

{ @param F a float }
{ @param G a float, greater than f }
{ @param Point some coordinates }
{ @param N a number }
{ @param FloatList a list of floats }
{ @param OtherList a list of floats }
{ @param Inlined some inlined structs }
{ @param Multiline a multiline struct }
procedure Floats(F: double; G: double; const Point: Coordinates; N: longint; const FloatList: T_FloatList; const OtherList: T_OtherList; const Inlined: T_Inlined; const Multiline: MultilineMix);
begin
    {* TODO Parsing is often easy, reprint mode is harder *}
end;

var
    F: double; { a float }
    G: double; { a float, greater than f }
    Point: Coordinates; { some coordinates }
    N: longint; { a number }
    FloatList: T_FloatList; { a list of floats }
    OtherList: T_OtherList; { a list of floats }
    Inlined: T_Inlined; { some inlined structs }
    Multiline: MultilineMix; { a multiline struct }
    i: longint;
    _: char;
begin
    readln(F);
    readln(G);
    readln(Point.X, Point.Y, Point.Z);
    readln(N);
    setLength(FloatList, N);
    for i := 0 to N - 1 do
        read(FloatList[i]);
    readln();
    setLength(OtherList, 9);
    for i := 0 to 9 - 1 do
        read(OtherList[i]);
    readln();
    setLength(Inlined, 3);
    for i := 0 to 3 - 1 do
    begin
        readln(Inlined[i].Integer_, _, Inlined[i].Char_, Inlined[i].Float);
    end;
    readln(Multiline.Integer2);
    readln(Multiline.String_);
    readln(Multiline.Float2);
    Floats(F, G, Point, N, FloatList, OtherList, Inlined, Multiline);
end.
